src/query/storages/fuse/fuse/src/operations/fuse_native_source.rs

```diff
@@ -0,0 +1,248 @@
+//  Copyright 2021 Datafuse Labs.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+use std::any::Any;
+use std::fs::File;
+use std::io::BufReader;
+use std::io::Read;
+use std::io::Take;
+use std::sync::Arc;
+
+use common_arrow::native::read::reader::PaReader;
+use common_arrow::native::read::PaReadBuf;
+use common_base::base::Progress;
+use common_base::base::ProgressValues;
+use common_catalog::plan::PartInfoPtr;
+use common_catalog::table_context::TableContext;
+use common_datablocks::DataBlock;
+use common_datavalues::ColumnRef;
+use common_datavalues::DataType;
+use common_exception::ErrorCode;
+use common_exception::Result;
+use common_functions::scalars::FunctionContext;
+use common_pipeline_core::processors::port::OutputPort;
+use common_pipeline_core::processors::processor::Event;
+use common_pipeline_core::processors::processor::ProcessorPtr;
+use common_pipeline_core::processors::Processor;
+use common_sql::evaluator::EvalNode;
+
+use crate::io::BlockReader;
+
+type DataChunks = Vec<(usize, PaReader<Box<dyn PaReadBuf + Send + Sync>>)>;
+
+enum State {
+    ReadData(Option<PartInfoPtr>),
+    Deserialize(DataChunks),
+    Generated(DataBlock, DataChunks),
+    Finish,
+}
+
+pub struct FuseNativeSource {
+    state: State,
+    ctx: Arc<dyn TableContext>,
+    scan_progress: Arc<Progress>,
+    output: Arc<OutputPort>,
+    output_reader: Arc<BlockReader>,
+
+    prewhere_reader: Arc<BlockReader>,
+    prewhere_filter: Arc<Option<EvalNode>>,
+    remain_reader: Arc<Option<BlockReader>>,
```

Comment:
If we don't pre-filter prewhere chunks, one reader is enough.

Reply:
Yes, I want to add `page` skip later.